### ISSUE TITLE
http IOCP BUG cause crashed 

### ISSUE THREAD

**Issue by wujvn71111** [libevent/libevent#957]
this is my demo

void http_req_gencb(struct evhttp_request* req, void* arg) {
	std::thread([req]() {
		struct evbuffer* buf = evbuffer_new();
		evbuffer_add_printf(buf, "<H1>Hello world!</H1>");
		evhttp_send_reply(req, HTTP_OK, "OK", buf);
		evbuffer_free(buf);
	}).detach();	
}
int main()
{
	WSADATA wsdata = {};
	WSAStartup(0x0202, &wsdata);
	evthread_use_windows_threads();
	struct event_config* cfg = event_config_new();
	event_config_set_flag(cfg, EVENT_BASE_FLAG_STARTUP_IOCP);
	event_config_set_num_cpus_hint(cfg, 8);
	struct event_base* base = event_base_new_with_config(cfg);
	event_config_free(cfg);
	struct evhttp* http = evhttp_new(base);
	evhttp_bind_socket(http, "0.0.0.0", 6790);
	evhttp_set_gencb(http, http_req_gencb, http);
	event_base_dispatch(base);
	event_base_free(base);
	WSACleanup();
	return 0;
}

**### when i send requset pipeline** ,the programe will be crashed
"[err] evhttp_read_cb: illegal connection state 7"

**Comment by wujvn71111** [comment]
clear EVENT_BASE_FLAG_STARTUP_IOCP ,the programme is running ok

**Comment by ploxiln** [comment]
Please use a "fenced code block" when pasting in code samples, so the formatting and special characters are not all messed up. Like so:

    ```
    // triple-backticks on their own line
    // both above and below
    int main() {
        printf("**code quote**\n");
        return 0;
    }
    ```

**Comment by wujvn71111** [comment]
#include <event2/event.h>
#include <event2/bufferevent.h>
#include <event2/thread.h>
#include <future>
#define GLOG_NO_ABBREVIATED_SEVERITIES
#include <glog/logging.h>
#pragma comment (lib, "ws2_32.lib")

void http_req_gencb(struct evhttp_request* req, void* arg) {
	auto uri = evhttp_request_get_evhttp_uri(req);	
	auto path=evhttp_uri_get_path(uri);
	auto inbuff = evhttp_request_get_input_buffer(req);
	auto sz = evbuffer_get_length(inbuff);
	auto data = evbuffer_pullup(inbuff, sz);
	LOG(INFO) << "rvc req:"<<path <<" content:"<<std::string((char*)data,sz);	
	std::thread([req]() {
		struct evbuffer* buf = evbuffer_new();
		evbuffer_add_printf(buf, "hello world");
		evhttp_send_reply(req, HTTP_OK, "OK", buf);
		evbuffer_free(buf);
	}).detach();	
}

int main()
{
	WSADATA wsdata = {};
	WSAStartup(0x0202, &wsdata);
	google::InitGoogleLogging("tt");
	FLAGS_alsologtostderr = true;
	evthread_use_windows_threads();
	struct event_config* cfg = event_config_new();
	event_config_set_flag(cfg, EVENT_BASE_FLAG_STARTUP_IOCP);
	event_config_set_num_cpus_hint(cfg, 8);
	struct event_base* base = event_base_new_with_config(cfg);
	event_config_free(cfg);
	struct evhttp* http = evhttp_new(base);
	evhttp_bind_socket(http, "0.0.0.0", 6790);
	evhttp_set_gencb(http, http_req_gencb, http);
	event_base_dispatch(base);
	event_base_free(base);
	WSACleanup();
	return 0;
}

**Comment by wujvn71111** [comment]
libevent version
#define EVENT__PACKAGE_VERSION "2.1.11"

**Comment by ploxiln** [comment]
You did not succeed at pasting your code in a code formatting block.

Anyway, it looks like the problem is that you are calling `evhttp_send_reply()` from a different thread than the event-loop thread. Even though you have enabled low-level cross-thread calls, the higher-level "bufferevent" which evhttp uses does not by default support cross-thread calls. (Mixing the event loop with threads is a somewhat advanced special case, usually you should keep all interaction with the event loop to a single fully event-based thread.)

This is pretty much the same question as #883 and the basic idea to support this is to use `evhttp_set_bevcb()` to hook into evhttp to create thread-safe bufferevents for it. Example from #883:

```diff
--- a/http-server.cpp
+++ b/http-server.cpp
@@ -23,6 +23,7 @@
 #include <event2/event.h>
 #include <event2/http.h>
 #include <event2/listener.h>
+#include <event2/bufferevent.h>
 #include <event2/buffer.h>
 #include <event2/util.h>
 #include <event2/keyvalq_struct.h>
@@ -86,6 +87,11 @@ void cb(evhttp_request *req, void *args)
 	ThreadPoolProvider::getPool()->addTask(task);  //run in threadpool
 }
 
+struct bufferevent* bev_cb(struct event_base *base, void *)
+{
+	return bufferevent_socket_new(base, -1, BEV_OPT_THREADSAFE);
+}
+
 int main(int argc, char **argv)
 {
 	evthread_use_pthreads();
@@ -124,6 +130,7 @@ int main(int argc, char **argv)
 		goto err;
 	}
 
+	evhttp_set_bevcb(http, bev_cb, NULL);
 	evhttp_set_gencb(http, cb, NULL);
 
 	if (event_base_dispatch(base) != 0)
```

**Comment by wujvn71111** [comment]
```
#include "pch.h"
#include <iostream>
#include <string>
#include <event.h>
#include <event2/http.h>
#include <event2/event.h>
#include <event2/bufferevent.h>
#include <event2/thread.h>
#include <future>
#define GLOG_NO_ABBREVIATED_SEVERITIES
#include <glog/logging.h>
#pragma comment (lib, "ws2_32.lib")

void http_req_gencb(struct evhttp_request* req, void* arg) {

	auto uri = evhttp_request_get_evhttp_uri(req);
	
	auto path=evhttp_uri_get_path(uri);
	auto inbuff = evhttp_request_get_input_buffer(req);
	auto sz = evbuffer_get_length(inbuff);
	auto data = evbuffer_pullup(inbuff, sz);
	LOG(INFO) << "rvc req:"<<path <<" content:"<<std::string((char*)data,sz);
	
	std::thread([req]() {
		struct evbuffer* buf = evbuffer_new();

		evbuffer_add_printf(buf, "<H1>Hello world!</H1>");
		evhttp_send_reply(req, HTTP_OK, "OK", buf);
		evbuffer_free(buf);
	}).detach();
	
}

struct bufferevent* bev_cb(struct event_base *base, void *)
{
	return bufferevent_socket_new(base, -1, BEV_OPT_THREADSAFE);
}


int main()
{
	WSADATA wsdata = {};
	WSAStartup(0x0202, &wsdata);
	google::InitGoogleLogging("tt");
	FLAGS_alsologtostderr = true;
	evthread_use_windows_threads();
	struct event_config* cfg = event_config_new();
	event_config_set_flag(cfg, EVENT_BASE_FLAG_STARTUP_IOCP);
	event_config_set_num_cpus_hint(cfg, 8);
	struct event_base* base = event_base_new_with_config(cfg);
	event_config_free(cfg);

	struct evhttp* http = evhttp_new(base);
	evhttp_bind_socket(http, "0.0.0.0", 6790);
	evhttp_set_bevcb(http, bev_cb, http);
	evhttp_set_gencb(http, http_req_gencb, http);

	event_base_dispatch(base);

	event_base_free(base);

	WSACleanup();
	return 0;
}
```
I add "bev_cb patch" ,but the program is still crashed

**Comment by wujvn71111** [comment]
I don't think the reason is "evbuffer" has no thread lock.
remove "event_config_set_flag(cfg, EVENT_BASE_FLAG_STARTUP_IOCP) ", it's  work fine.
I read evhttp_send_reply()  src code.
in evbuffer_add_buffer () has code  "EVBUFFER_LOCK2(inbuf, outbuf);"

**Comment by ploxiln** [comment]
I see, thanks for the update. I don't have experience with IOCP (just a few unixes), but someone with some IOCP experience will probably be able to comment here in a couple days.

**Comment by azat** [comment]
I'm seeing that you are using reply-from-thread, do you really need this?
There are some issues with this, I bet it will work okay if you will remove it.

But if you need this (for example if callback is really slow) you can run the program under `valgrind --tool=drd`/`valgrind --tool=helgrind` **or** compile with `-fsanitize=thread`/`-fsanitize=address` and post the errors that will be reported I will take a look

>in evbuffer_add_buffer () has code "EVBUFFER_LOCK2(inbuf, outbuf);"

This locks the buffer only if the lock is set.
For regular bufferevent's (`bufferevent_socket_new`) you need to enable thread-safety manually, while for IOCP they are enabled by default (doe to the "async" nature of IOCP).

>i think http running in iocp ,the http state is missing.

This is not an error, it is done by intention.

**Comment by wujvn71111** [comment]
sorry，I try to install valgrind on windows 10 ,but valgrind hasn't binary package for windows 10 .
I try to build by source but failed for "configure: error: Valgrind is operating system specific. Sorry. Please consider doing a port.
".



wujvn@sina.com
 
From: Azat Khuzhin
Date: 2020-02-10 18:28
To: libevent/libevent
CC: wujvn71111; Author
Subject: Re: [libevent/libevent] http IOCP BUG cause crashed (#957)
I'm seeing that you are using reply-from-thread, do you really need this?
There are some issues with this, I bet it will work okay if you will remove it.
But if you need this (for example if callback is really slow) you can run the program under valgrind --tool=drd/valgrind --tool=helgrind or compile with -fsanitize=thread/-fsanitize=address and post the errors that will be reported I will take a look
in evbuffer_add_buffer () has code "EVBUFFER_LOCK2(inbuf, outbuf);"
This locks the buffer only if the lock is set.
For regular bufferevent's (bufferevent_socket_new) you need to enable thread-safety manually, while for IOCP they are enabled by default (doe to the "async" nature of IOCP).
i think http running in iocp ,the http state is missing.
This is not an error, it is done by intention.

**Comment by azat** [comment]
Actually looking into this again and I don't think that valgrind or sanitizers will help.

@ygj6 can debug this problem on windows?

**Comment by ygj6** [comment]
I ran your program with or without the patch mentioned by *ploxiln*, and they all completed successfully. My Visual Studio 2019 has the following configuration:
`Windows SDK version: 10.0.18362.0`
`Platform Toolset: Visual Studio 2019 (v142)`
So, @wujvn71111,  what's your configuration of Visual Studio?

I also tried it with Visual Studio 2017 and got ok.

**Comment by wujvn71111** [comment]
http requests must send "pipeline".
Normal http request can't cause the crahe.
My test program simulates a hacker attack by http.
sending http requests pipeline in one connecton  without waiting reply.
libevent will crashed at IOCP configed,but remove IOCP it's work fine.
so i think something wrong in http with iocp.
 


wujvn@sina.com
 
From: yuangongji
Date: 2020-02-12 12:43
To: libevent/libevent
CC: wujvn71111; Mention
Subject: Re: [libevent/libevent] http IOCP BUG cause crashed (#957)
I ran your program with or without the patch mentioned by ploxiln, and they all completed successfully. My Visual Studio 2019 has the following configuration:
Windows SDK version: 10.0.18362.0
Platform Toolset: Visual Studio 2019 (v142)
So, @wujvn71111, what's your configuration of Visual Studio?
I also tried it with Visual Studio 2017 and got ok.

**Comment by ygj6** [comment]
I will test pipeline requet to evhttp.
Is your code confidential?  Can you push the test code to this [repo](https://github.com/ygj6/libevent-http-server) or any other repository you like?

**Comment by ygj6** [comment]
I wrote a pipeline client via `libcurl` whose version from 7.43.0 to 7.61.0 supports pipeline , you can see the results [here](https://github.com/ygj6/libevent-http-server/runs/472486748?check_suite_focus=true). Can you modify this code and reproduce the crash?

**Comment by wujvn71111** [comment]
this is my test code here ,
```package main

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"text/template"
)

var req = `GET / HTTP/1.1
Host:{{.}}
Accept: image/gif, image/jpeg, */*
Accept-Language: en-us
Accept-Encoding: gzip, deflate
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)

`

func main() {
	addr := os.Args[1]
	buf := new(bytes.Buffer)
	tmp := template.New("hello")
	tmp, _ = tmp.Parse(req)
	tmp.Execute(buf, addr)
	fmt.Println(string(buf.Bytes()))
	con, err := net.Dial("tcp", addr)

	if err != nil {
		log.Fatalf("dail %v err:%v", addr, err.Error())
	}

	for i := 0; i < 100; i++ {
		_, err = con.Write(buf.Bytes())
		if err != nil {
			log.Fatalf("write  err:%v", err.Error())
		}
	}

	_, err = io.Copy(os.Stdout, con)
	if err != nil {
		log.Fatalf("read  err:%v", err.Error())
	}
}
```
this is the crash promptI0316 12:26:22.213692  3140 testevhttpsvr.cpp:117] start http server
I0316 12:26:40.806047 25100 testevhttpsvr.cpp:43] rvc req:/ content:
[err] evhttp_read_cb: illegal connection state 7




wujvn@sina.com
 
From: yuangongji
Date: 2020-02-27 22:04
To: libevent/libevent
CC: wujvn71111; Mention
Subject: Re: [libevent/libevent] http IOCP BUG cause crashed (#957)
I wrote a pipeline client via libcurl whose version from 7.43.0 to 7.61.0 supports pipeline , you can see the results here. Can you modify this code and reproduce the crash?

**Comment by wujvn71111** [comment]
this is the exe file 
you can test at cmd window
"PipelineTest 127.0.0.1:6790"



wujvn@sina.com
 
From: wujvn@sina.com
Date: 2020-03-16 12:28
To: libevent/libevent; libevent/libevent
CC: Mention
Subject: Re: Re: [libevent/libevent] http IOCP BUG cause crashed (#957)
this is my test code here ,
```package main

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"text/template"
)

var req = `GET / HTTP/1.1
Host:{{.}}
Accept: image/gif, image/jpeg, */*
Accept-Language: en-us
Accept-Encoding: gzip, deflate
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)

`

func main() {
	addr := os.Args[1]
	buf := new(bytes.Buffer)
	tmp := template.New("hello")
	tmp, _ = tmp.Parse(req)
	tmp.Execute(buf, addr)
	fmt.Println(string(buf.Bytes()))
	con, err := net.Dial("tcp", addr)

	if err != nil {
		log.Fatalf("dail %v err:%v", addr, err.Error())
	}

	for i := 0; i < 100; i++ {
		_, err = con.Write(buf.Bytes())
		if err != nil {
			log.Fatalf("write  err:%v", err.Error())
		}
	}

	_, err = io.Copy(os.Stdout, con)
	if err != nil {
		log.Fatalf("read  err:%v", err.Error())
	}
}
```
this is the crash promptI0316 12:26:22.213692  3140 testevhttpsvr.cpp:117] start http server
I0316 12:26:40.806047 25100 testevhttpsvr.cpp:43] rvc req:/ content:
[err] evhttp_read_cb: illegal connection state 7




wujvn@sina.com
 
From: yuangongji
Date: 2020-02-27 22:04
To: libevent/libevent
CC: wujvn71111; Mention
Subject: Re: [libevent/libevent] http IOCP BUG cause crashed (#957)
I wrote a pipeline client via libcurl whose version from 7.43.0 to 7.61.0 supports pipeline , you can see the results here. Can you modify this code and reproduce the crash?

**Comment by ygj6** [comment]
Thanks for you providing the test code. 
I may not be able to reply to you soon, because I need to set up a golang development and debug environment first.
